<a href="https://colab.research.google.com/github/gbodra/mestrado-ia-pel208-deep-learning/blob/main/DeepLearning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import tensorflow as tf
from tensorflow import keras
import numpy as np
import time
import os

(train_x, train_y), (test_x, test_y) = keras.datasets.mnist.load_data()
train_x = train_x / 255.0
test_x = test_x / 255.0

train_x = tf.expand_dims(train_x, 3)
test_x = tf.expand_dims(test_x, 3)

val_x = train_x[:5000]
val_y = train_y[:5000]

11493376/11490434 [==============================] - 0s 0us/step


In [5]:
lenet_5_model = keras.models.Sequential([
    keras.layers.Conv2D(6, kernel_size=5, strides=1,  activation='tanh', input_shape=train_x[0].shape, padding='same'),
    keras.layers.AveragePooling2D(),
    keras.layers.Conv2D(16, kernel_size=5, strides=1, activation='tanh', padding='valid'),
    keras.layers.AveragePooling2D(),
    keras.layers.Flatten(),
    keras.layers.Dense(120, activation='tanh'),
    keras.layers.Dense(84, activation='tanh'),
    keras.layers.Dense(10, activation='softmax')
])

In [6]:
lenet_5_model.compile(optimizer='adam', loss=keras.losses.sparse_categorical_crossentropy, metrics=['accuracy'])

In [7]:
lenet_5_model.fit(train_x, train_y, epochs=100, validation_data=(val_x, val_y))

Epoch 1/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.2266 - accuracy: 0.9329 - val_loss: 0.0881 - val_accuracy: 0.9746
Epoch 2/10
1875/1875 [==============================] - 41s 22ms/step - loss: 0.0886 - accuracy: 0.9724 - val_loss: 0.0541 - val_accuracy: 0.9850
Epoch 3/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.0611 - accuracy: 0.9816 - val_loss: 0.0360 - val_accuracy: 0.9900
Epoch 4/10
1875/1875 [==============================] - 42s 22ms/step - loss: 0.0474 - accuracy: 0.9850 - val_loss: 0.0281 - val_accuracy: 0.9900
Epoch 5/10
1875/1875 [==============================] - 41s 22ms/step - loss: 0.0370 - accuracy: 0.9883 - val_loss: 0.0227 - val_accuracy: 0.9934
Epoch 6/10
1875/1875 [==============================] - 42s 22ms/step - loss: 0.0315 - accuracy: 0.9899 - val_loss: 0.0235 - val_accuracy: 0.9928
Epoch 7/10
1875/1875 [==============================] - 42s 22ms/step - loss: 0.0268 - accuracy: 0.9915 - val_loss: 0.0159 -

In [8]:
lenet_5_model.evaluate(test_x, test_y)

313/313 [==============================] - 4s 12ms/step - loss: 0.0497 - accuracy: 0.9855


[0.04965367913246155, 0.9854999780654907]

# Análise dos resultados

## LeNet-5

| Epochs | Tempo de treinamento | Perda | Acuracidade |
| ------ | -------------------- | ----- | ----------- |
| 5      | 206s                 | 0.0567| 0.9810      |
| 10     | 416s                 | 0.0497| 0.9855      |
| 100